storage/storage_dest.go

```diff
@@ -117,18 +117,18 @@ func newImageDestination(sys *types.SystemContext, imageRef storageReference) (*
 			HasThreadSafePutBlob:           true,
 		}),
 
-		imageRef:              imageRef,
-		directory:             directory,
-		signatureses:          make(map[digest.Digest][]byte),
-		blobDiffIDs:           make(map[digest.Digest]digest.Digest),
-		blobAdditionalLayer:   make(map[digest.Digest]storage.AdditionalLayer),
-		fileSizes:             make(map[digest.Digest]int64),
-		filenames:             make(map[digest.Digest]string),
-		SignatureSizes:        []int{},
-		SignaturesSizes:       make(map[digest.Digest][]int),
-		indexToStorageID:      make(map[int]*string),
-		indexToAddedLayerInfo: make(map[int]addedLayerInfo),
-		diffOutputs:           make(map[digest.Digest]*graphdriver.DriverWithDifferOutput),
+		imageRef:                imageRef,
+		directory:               directory,
+		signatureses:            make(map[digest.Digest][]byte),
+		uncompressedOrTocDigest: make(map[digest.Digest]digest.Digest),
+		blobAdditionalLayer:     make(map[digest.Digest]storage.AdditionalLayer),
+		fileSizes:               make(map[digest.Digest]int64),
+		filenames:               make(map[digest.Digest]string),
```

Comment:
AFAICS all of this is no longer correct, because it identifies layers by (compressed) total blob digest, but we don’t always use or validate that value. Renaming `blobDiffIDs` to `uncompressedOrTocDigest` documents the cause but isn’t sufficient to make the data reliable. (E.g. consider the case of a manifest with
```
[
 layer1: { digest = sharedDigest } // found as an additional layer / as existing layer / not found and stored in .filenames
 layer2: { digest = sharedDigest, TOCDigest = otherTOCDigest }
 layer3: { digest = otherTotalDigest, TOCDigest = sharedDigest }
]
```
it gets too hard to understand what the keys of the maps mean.)

So I think `tryReusingBlobAsPending`/`putBlobToPendingFile` and `commitLayer` should coordinate a bit more closely, and _primarily_ coordinate by layer index.

Something along these lines, maybe?
- In addition, to `indexTo{Added,Committed}LayerInfo`, maintain `indexTo{Pending,NotNecessarilyCommitted,Known}LayerInfo`, so that `.Commit()` can call `commitLayer` with a full `addedLayerInfo` struct
- In `addedLayerInfo`, track:
  - `individualLayerID`: a string or `[]byte`, not a `digest.Digest`; used to construct the “chain” layer ID in `commitLayer`; currently these are the `blobDiffIDs` values, but if we track them in `addedLayerInfo`, we can actually make them unambiguous: Use DiffID _if it is known_, and use a distinct namespace (`\TOC:…` ?) for layers identified by TOC digest.
    - That would clearly disambiguate TOC-digest layers vs. and total-digest layers, ensure that the `lastLayer` values used creating/deduplicating images account for these distinctions, and help us compute a distinct image ID.
  - Where we found the layer data = what to commit if the layer chain ID does not exist locally. A minimal version of that would just be “use DiffID x (with existing code)” vs. “use TOC y (with new code)”, but we could probably, and cleanly, directly record more of the data found by `tryReusingBlobAsPending` directly into `addedLayerInfo`.
    - That would mostly replace `blobAdditionalLayer`/`diffOutputs`; these are ~write-only apart from clean-up, so we don’t need to index them by total blob ID.
    -  For layers found by DiffID , we could record the layer ID instead of just the DiffID, to avoid unnecessary lookups, similarly to the new `storageImageSource` code.
    - (We would still maintain the existing `filenames`/`fileSizes` for the rare case of images that contain the same locally-nonexistent layer twice.)

Reply:
Maybe this refactoring would be better to do separately, as a prerequisite smaller PR.

Reply:
The correctness issue was ultimately resolved in #2288 , in a less disruptive way.